docs: add mrvi user guide (#2761)
scverse/scvi-tools: docs/user_guide/models/mrvi.md

# MrVI

**MrVI** [^ref1] (Multi-resolution Variational Inference; Python class
{class}`~scvi.external.MRVI`) is a deep generative model designed for the analysis of large-scale
single-cell transcriptomics data with multi-sample, multi-batch experimental designs.

The advantages of MrVI are:

> - Facilitates both exploratory analysis to divide samples into groups based on cellular and

Reviewer: What's cellular and molecular? Can we just say molecular?

    molecular properties and comparative analysis to compare pre-defined groups of samples.
- Provides single-cell-resolution differential expression and differential abundance estimates.
- Captures nonlinear, cell-type specific effects of sample-level covariates on gene expression.

> - Scales to large datasets with millions of cells and hundreds of samples.

Reviewer: Is this unique over other models?

Reviewer 2: no (mostly because the only fair benchmarks are the composition ones), but i'd still like to leave it since by omission it implies we arent scalable (same note is in the other user guides).

The limitations of MrVI include:

> - Requires specification of which sample-level covariates are of interest vs nuisance factors.

Reviewer: Target covariate?

- Differential expression procedure does not provide single-cell gene-level p-values.

```{topic} Tutorials:

-    {doc}`tutorials/notebooks/scrna/MrVI_tutorial`
```

## Preliminaries

MrVI takes as input a scRNA-seq gene expression matrix $X$ with $N$ cells and $G$ genes.
Additionally, it requires specification of:

- A sample ID $s_n$ for each cell $n$, denoting which sample the cell originated from,
- Nuisance covariates $b_n$ for each cell $n$ (e.g. sequencing run, processing day),

> - (Optional) Additional sample-level covariates of interest $c_s$ for each sample $s$ (e.g.

Reviewer: Add here optional celltype labels?

# MrVI

**MrVI** [^ref1] (Multi-resolution Variational Inference; Python class
{class}`~scvi.external.MRVI`) is a deep generative model designed for the analysis of large-scale
single-cell transcriptomics data with multi-sample, multi-batch experimental designs.

The advantages of MrVI are:

> 
> - Facilitates both exploratory analysis to divide samples into groups based on
>     molecular properties and comparative analysis to compare pre-defined groups of samples.
> - Provides single-cell-resolution differential expression and differential abundance estimates.
> - Captures nonlinear, cell-type specific effects of sample-level covariates on gene expression.
> - Scales to large datasets with millions of cells and hundreds of samples.
> 

Reviewer 3: ```suggestion
MrVI conducts both **exploratory analyses** (locally dividing samples into groups based on molecular properties)
and **comparative analyses** (comparing pre-defined groups of samples in terms of differential expression and differential abundance) at single-cell resolution.
It can capture nonlinear and cell-type specific variation of sample-level covariates on gene expression.

For this purpose, MrVI requires to specify (1) a sample-level target covariate of interest (that generally corresponds to sample IDs) and (2) nuisance covariates (e.g., sequencing batch, processing day) that introduce technical variation.

```

The limitations of MrVI include:

- Requires specification of sample-level target (i.e., of biological interest) and nuisance (i.e., introduces undesired variation) covariates.
- Differential expression procedure does not provide single-cell gene-level p-values.

> 

Reviewer 3: ```suggestion
MrVI conducts both **exploratory analyses** (locally dividing samples into groups based on molecular properties)
and **comparative analyses** (comparing pre-defined groups of samples in terms of differential expression and differential abundance) at single-cell resolution.
It can capture nonlinear and cell-type specific variation of sample-level covariates on gene expression.
```

```{topic} Tutorials:

-    {doc}`tutorials/notebooks/scrna/MrVI_tutorial`
```

## Preliminaries

MrVI takes as input a scRNA-seq gene expression matrix $X$ with $N$ cells and $G$ genes.
Additionally, it requires specification of:

> 

Reviewer 3: ```suggestion
Additionally, it requires specification, for each cell $n$ of
```

     - Mixture of Gaussians means for prior on $u_n$.
     - ``u_prior_means``
   * - :math:`\Sigma_1, ..., \Sigma_K \in \mathbb{R}^{L \times L}`
     - Mixture of Gaussians covariance matrices for prior on $u_n$.
     - ``u_prior_scales``
   * - :math:`\pi_1, ..., \pi_K \in \mathbb{R}^+`
     - Mixture of Gaussians weights for prior on $u_n$.
     - ``u_prior_logits``
```

## Inference

MrVI uses variational inference to approximate the posterior of $u_n$ and $z_n$. The variational
distributions are:

$q_{\phi}(u_n | x_n) := \mathcal{N}(\mu_{\phi}(x_n), \sigma^2_{\phi}(x_n)I)$

$z_n := u_n + f_{\phi}(u_n, s_n)$

Here $\mu_{\phi}, \sigma^2_{\phi}$ are encoder neural networks and $f_{\phi}$ is a deterministic
mapping based on multi-head attention between $u_n$ and a learned embedding for sample $s_n$.

## Tasks

### Exploratory analysis

MrVI enables unsupervised local sample stratification via the construction of cell-specific
sample-sample distance matrices:

1. For any cell state $u$, compute counterfactual cell states $z^{(s)}$ for each sample $s$
2. Compute sample-sample distance matrices $D^{(n)}$ based on the Euclidean distance between the
    $z^{(s)}$ for each cell $n$

> 3. Cluster cells based on their $D^{(n)}$ to find cell populations with distinct sample

Reviewer 3: ```suggestion
1. For each cell state $u_n$ (cell $n$), compute counterfactual cell states $z^{(s)}_n$ for all possible samples $s$
2. Compute cell-specific sample-sample distance matrices $D^{(n)}$ based on the Euclidean distance between all pairs of $z^{(s)}_n$
3. Cluster cells based on their $D^{(n)}$ to find cell populations with distinct sample stratifications
```

# MrVI

**MrVI** [^ref1] (Multi-resolution Variational Inference; Python class
{class}`~scvi.external.MRVI`) is a deep generative model designed for the analysis of large-scale
single-cell transcriptomics data with multi-sample, multi-batch experimental designs.

> The advantages of MrVI are:

Reviewer 3: I would remove this altogether

> The advantages of MrVI are:

Reviewer 3: Sorry I have issues with suggestions, but I would remove all of this

```
The advantages of MrVI are:

 - Facilitates both exploratory analysis to divide samples into groups based on
     molecular properties and comparative analysis to compare pre-defined groups of samples.
 - Provides single-cell-resolution differential expression and differential abundance estimates.
 - Captures nonlinear, cell-type specific effects of sample-level covariates on gene expression.
 - Scales to large datasets with millions of cells and hundreds of samples.

 The limitations of MrVI include:

 - Requires specification of sample-level target (i.e., of biological interest) and nuisance (i.e., introduces undesired variation) covariates.
 - Differential expression procedure does not provide single-cell gene-level p-values.
```

     - Mixture of Gaussians covariance matrices for prior on $u_n$.
     - ``u_prior_scales``
   * - :math:`\pi_1, ..., \pi_K \in \mathbb{R}^+`
     - Mixture of Gaussians weights for prior on $u_n$.
     - ``u_prior_logits``
```

## Inference

MrVI uses variational inference to approximate the posterior of $u_n$ and $z_n$. The variational
distributions are:

$q_{\phi}(u_n | x_n) := \mathcal{N}(\mu_{\phi}(x_n), \sigma^2_{\phi}(x_n)I)$

$z_n := u_n + f_{\phi}(u_n, s_n)$

Here $\mu_{\phi}, \sigma^2_{\phi}$ are encoder neural networks and $f_{\phi}$ is a deterministic
mapping based on multi-head attention between $u_n$ and a learned embedding for sample $s_n$.

## Tasks

### Exploratory analysis

MrVI enables unsupervised local sample stratification via the construction of cell-specific
sample-sample distance matrices, for every cell $n$:

1. For each cell state $u_n$, compute counterfactual cell states $z^{(s)}_n$ for all possible samples $s$.
2. Compute cell-specific sample-sample distance matrices $D^{(n)}$ based on the Euclidean distance between all pairs of $z^{(s)}_n$.
3. Cluster cells based on their $D^{(n)}$ to find cell populations with distinct sample stratifications.

> 4. Average $D^{(n)}$ within each cell cluster and hierarchically cluster samples

Reviewer 3: ```suggestion
MrVI also enables supervised comparative analysis to detect cell-type specific DE and DA between sample groups.

#### Differential expression
At a high level, the DE procedure regresses, within each cell $n$, counterfactual cell states $z^{(s)}_n$ on sample-level covariates $c_s$ of interest for analysis as
$z^{(s)}_n = \beta_n c_s + \beta_0 + \epsilon_n$.
For instance, if $c_s$ is a binary covariate, then $\beta_n$ will capture the shift (in $z$-space) induced by samples for which $c_s = 1$ compared to samples for which $c_s = 0$.

This procedure, repeated for all cells, allows several downstream analyses.
First, comparing the norm of $\beta_n$ (using $\chi^2$ statistics) across cells can identify cell-states that vary the most for a given covariate, or conversely, identify sample covariates that strongly associate with specific cell states.

Second, by decoding the linear approximation of $z^{(s)}_n$ for different covariate vectors that we would like to compare, we can compute associated log fold-changes to identify DE genes at the cell level.

```

This automatically reveals distinct sample stratifications that are specific to particular cell
subsets.

### Comparative analysis

MrVI also enables supervised comparative analysis to detect cell-type specific DE and DA between
sample groups:

> 

Reviewer 3: ```suggestion
MrVI also enables supervised comparative analysis to detect cell-type specific DE and DA between sample groups.

#### Differential expression
At a high level, the DE procedure regresses, within each cell $n$, counterfactual cell states $z^{(s)}_n$ on sample-level covariates $c_s$ of interest for analysis as
$z^{(s)}_n = \beta_n c_s + \beta_0 + \epsilon_n$.
For instance, if $c_s$ is a binary covariate, then $\beta_n$ will capture the shift (in $z$-space) induced by samples for which $c_s = 1$ compared to samples for which $c_s = 0$.

This procedure, repeated for all cells, allows several downstream analyses.
First, comparing the norm of $\beta_n$ (using $\chi^2$ statistics) across cells can identify cell-states that vary the most for a given covariate, or conversely, identify sample covariates that strongly associate with specific cell states.

Second, by decoding the linear approximation of $z^{(s)}_n$ for different covariate vectors that we would like to compare, we can compute associated log fold-changes to identify DE genes at the cell level.

#### Differential abundance
```

The differential expression procedure at a high level is:

> 

Reviewer 3: ```suggestion
To compare two sets of samples, MrVI computes the log-ratio between the aggregated posteriors of two groups, $S_1$ and $S_2$.

```

1. For each cell $n$, regress $z^{(s)}_n$ on sample-level covariates $c_s$ to obtain cell-level
    coefficients $\beta_n$,
2. Identify significant covariates based on $\chi^2$ statistic of $\beta_n$,
3. Predict and decode $z^{(s')}_n$ for given covariate vectors and compute log fold-changes to get
    cell-level DE genes.

The differential abundance procedure at a high level is:

> 

Reviewer 3: ```suggestion
This aggregated posterior $q_s$ characterizes the distribution of all cells in sample $s$.
To characterize the distribution of cells in a group of samples $S$, we can consider the mixture of aggregated posteriors $q_s$ for all $s \in S$, corresponding to 
$q_S := \frac{1}{|S|} \sum_{s \in S} q_s$.
In particular, cell states $u$ that are abundant in $S$ will have a high probability mass in $q_S$, while rare states will have low probability mass.
```

> 

Reviewer 3: ```suggestion
### Comparative analysis

MrVI also enables supervised comparative analysis to detect cell-type specific DE and DA between sample groups.

#### Differential expression
At a high level, the DE procedure regresses, within each cell $n$, counterfactual cell states $z^{(s)}_n$ on sample-level covariates $c_s$ of interest for analysis as
$z^{(s)}_n = \beta_n c_s + \beta_0 + \epsilon_n$.
For instance, if $c_s$ is a binary covariate, then $\beta_n$ will capture the shift (in $z$-space) induced by samples for which $c_s = 1$ compared to samples for which $c_s = 0$.

This procedure, repeated for all cells, allows several downstream analyses.
First, comparing the norm of $\beta_n$ (using $\chi^2$ statistics) across cells can identify cell-states that vary the most for a given covariate, or conversely, identify sample covariates that strongly associate with specific cell states.

Second, by decoding the linear approximation of $z^{(s)}_n$ for different covariate vectors that we would like to compare, we can compute associated log fold-changes to identify DE genes at the cell level.

#### Differential abundance

To compare two sets of samples, MrVI computes the log-ratio between the aggregated posteriors of two groups, $S_1$ and $S_2$.

In particular, the aggregated posterior for any sample $s$ is defined as
$q_s := \frac{1}{|s|} \sum_{n, s_n=s} q^{u}_{n}$,
where $q_n$ is the posterior of cell $n$ in $u$-space.

This aggregated posterior $q_s$ characterizes the distribution of all cells in sample $s$.
To characterize the distribution of cells in a group of samples $S$, we can consider the mixture of aggregated posteriors $q_s$ for all $s \in S$, corresponding to 
$q_S := \frac{1}{|S|} \sum_{s \in S} q_s$.
In particular, cell states $u$ that are abundant in $S$ will have a high probability mass in $q_S$, while rare states will have low probability mass.

More generally, we can consider the log-ratio of aggregated posteriors between two groups of samples $S_1$ and $S_2$ as a measure of differential abundance:
$r = \log \frac{q_{S_1}}{q_{S_2}}$.
We can evaluate these log-ratios for all cell states $u$ to identify DA cell-state regions.
In particular, large positive (resp. negative) values of $r$ indicate that cell states are more abundant in $S_1$ (resp. $S_2$).
```
